### PR TITLE
Allow specifying similarity function for sentence-similarity task

### PR DESCRIPTION
**Pull request by osanseviero**

### CODE REVIEW

File: api-inference-community/tests/test_nlp.py [hunk 1]
```diff
@@ -56,6 +56,33 @@ def test_missing_input(self):
...
+            "similarity_method": "dot_product",
+        }
+        bpayload = json.dumps({"inputs": inputs}).encode("utf-8")
+        normalized_inputs, processed_params = normalize_payload_nlp(
+            bpayload, "sentence-similarity"
+        )
+        self.assertEqual(processed_params, {})
+        self.assertEqual(inputs, normalized_inputs)
+
+    def test_invalid_similarity_function(self):
```
Comment (Narsil) [inline api-inference-community/tests/test_nlp.py:74]: We're missing a test for default behavior at least.

File: api-inference-community/api_inference_community/validation.py [hunk 2]
```diff
@@ -114,9 +115,18 @@ class QuestionInputsCheck(BaseModel):
     context: str
 
 
+class SimilarityFunctionEnum(str, Enum):
+    default = "default"
```
Comment (Narsil) [inline api-inference-community/api_inference_community/validation.py:119]: What is the default ? Isn't it `cosine_similarity`? 

Or maybe it should just be `None` if we can't force it.  `default` is always an issue, because it's unclear what it is.

I am not aware of any model than can be used with `both` or `either` scheme, so I feel it should be the framework's responsability to actually handle this (hence, NOT being a parameter, just like the tokenization is not a choice of the pipeline)

**Review by Narsil — CHANGES_REQUESTED** [review]
I think it should be the framework's responsibility to handle the similarity function as it is tightly linked (AFAIK) to the model, hence it does not make sense to use a difference similarity on a given model.

If it's not true then the `default` should be handled better. (The fact that the `default` is a bit blurry, is a tell IMO that something is wrong)

**Comment by osanseviero** [comment]
I'm checking with Nils to see if this could somehow be in the model configuration. In theory the similarity function could be in the configuration json and then it would completely be on the framework side, as you suggest.

**Comment by nreimers** [comment]
Hi,
Yes, it would make sense to have the similarity function be specified in the config of the model. Otherwise users have to guess which similarity function is suitable and often choose the wrong one, which gives inferior results.

I think this would require a new file that is stored along with the model that stores this information. In the version < 2 there was a config.json for sentence-transformers, but it was removed to version so that it can be hosted on the hub. We might need to re-introduce a new file like 'sentence-transformers-config.json' which can host such information.

**Comment by osanseviero** [comment]
Alternatively, could we combine `modules.json` with this?

Everything related to the config would go in a single `json`

```
{
  similarity: `dot_score`,
  modules: [
    {
      "idx": 0,
      "name": "0",
      "path": "",
      "type": "sentence_transformers.models.Transformer"
    },
    {
      "idx": 1,
      "name": "1",
      "path": "1_WKPooling",
      "type": "sentence_transformers.models.WKPooling"
    }
  ]
```

**Comment by nreimers** [comment]
Yes, this would be an option too. The code would need to be updated that it could recognize modules.json with the old format.

Also I think how we can set the similarity parameter:
- There should be a method, where you can change it by hand (e.g. model.set_default_similarity('dot_product'))
- Which similarity function can be used is quite complex:

1) It depends how you train the model. There are some loss functions, that optimize the model for either cosine similarity or for dot product (you specify this as a parameter for `losses.MultipleNegativeRankingLoss`)
2) Some people train models that produce embeddings with unit length. In that case, cosine similarity and dot_product are identical and both can be used. dot_product is in that case computationally more efficient, i.e. would be better to use. But it would be good to know if the model can also be used with cosine_similarity, as some ANN methods don't support dot-product with unnormalized vectors.
3) Some models are also compatible with multiple similarity functions, i.e. you get the same results independent if you use cosine_similartiy, manhatten_distance, or euclidean_distance => Knowing this can be relevant when you do clustering with e.g. AgglomerativeClustering from Sklearn, as it supports better methods for euclidean_distance than for e.g. cosine similarity.

**Comment by Narsil** [comment]
> 
>     3. Some models are also compatible with multiple similarity functions, i.e. you get the same results independent if you use cosine_similartiy, manhatten_distance, or euclidean_distance => Knowing this can be relevant when you do clustering with e.g. AgglomerativeClustering from Sklearn, as it supports better methods for euclidean_distance than for e.g. cosine similarity.

Interesting, TIL !
But if all are equivalent, is there a single obvious best choice, or not ? Seems like euclidean is better/simpler all the time when equivalent to cosine for instance

**Comment by nreimers** [comment]
Hi @Narsil
there is sadly not a single, best choice.

There can differences, e.g.:
- The computational complexity is different. Dot-product is really fast, especially on GPUs (just a matrix multiplication). Cosine-similarity is also fast, especially often you can normalize vectors and then cosine similarity is just a dot-product. Not sure how fast euclidean distance can be computed for unnormalized vectors (for normalized vectors it is just eucl(A,B) = 2(1-cossim(A,B)) )

- There can be restrictions from external libraries: E.g. AgglomerativeClustering from Sklearn works best with ward linkage, which requires euclidean distance. The semantic search engine Vespa.ai works better with cosine similarity, using it with dot product is really really complicated. There are some other ANN libraries, where you need to do extra work to use specific similarity functions.


The most common similarity functions are cosine similarity and dot-product, especially for semantic search. But clustering algorithms often work implicit or explicitly with euclidean distance (like k-means is based on euclidean distance).

So it would be good to know which similarity functions are compatible with a model. This gives the (power) user the choice to use it correctly / to avoid complicated vector space conversions to make it compatible with a certain similarity function.
